### PR TITLE
Add leader-elected shard sync on application bootstrap

### PR DESCRIPTION
**Pull request by micah-jaffe**
Initial commit for leader elected shard sync on application bootstrap for KCL 1.x. Still need to add unit tests for new functionality and implement periodic shard sync for ShardEndShardSync strategy. Existing unit tests pass.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java
Comment (micah-jaffe) [inline src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java]: Changed these while debugging unit tests, will change back to commented out values in next revision.

File: src/test/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/WorkerTest.java
Comment (micah-jaffe) [inline src/test/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/WorkerTest.java]: Forgot to remove this from local testing as well - will remove on next revision

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java
Comment (joshua-kim) [reply src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java]: For future reference, you can always use `git add -p` to partially stage changes

File: src/test/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/WorkerTest.java
Comment (joshua-kim) [inline src/test/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/WorkerTest.java]: We shouldn't need this I believe, we will always be using getShards with filter in the empty case. (L450)

File: src/test/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/WorkerTest.java
Comment (joshua-kim) [inline src/test/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/WorkerTest.java:439]: Related to above, we only need this mocked call I believe.

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/PeriodicShardSyncManager.java
Comment (joshua-kim) [inline src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/PeriodicShardSyncManager.java]: Why do we not propagate any exception to TaskResult?

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java
Comment (joshua-kim) [inline src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java:697]: I believe this code path will never be executed since we never propagate the exception initially

**Review by joshua-kim — CHANGES_REQUESTED** [review]

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/PeriodicShardSyncManager.java
Comment (micah-jaffe) [reply src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/PeriodicShardSyncManager.java]: Any exception that occurs in `metricsEmittingShardSyncTask.call()` will be caught and propagated in `result`, which holds the last exception as a member. The only case where the Throwable catch block would be exercised is if something goes wrong in the logging or the `isLeader` method, neither of which are expected to throw. This was kept mostly for consistency with existing code.

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java
Comment (micah-jaffe) [reply src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java]: Good to know!

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java
Comment (micah-jaffe) [reply src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java:697]: The exception gets propagated within the TaskResult in metricsEmittingShardSyncTask.call() (TaskResult) -> runShardSync (TaskResult) -> syncShardsOnce (TaskResult). So any exception in the shard sync task will make it here. I'll add a unit test to validate.

File: src/test/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/WorkerTest.java
Comment (micah-jaffe) [reply src/test/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/WorkerTest.java]: Will remove this line

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java
Comment (micah-jaffe) [reply src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java:697]: Actually, this behavior is already validated in `testInitializationFailuresWithRetries()`. KinesisProxy is mocked to throw on calling the service, and KCL retries initialization attempts, meaning this code was exercised.

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/PeriodicShardSyncManager.java
Comment (ychunxue) [inline src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/PeriodicShardSyncManager.java]: In runShardSync method, the PeriodicShardSyncManager will elect leaders and may not run shardSyncTask. This means shardSync is not guaranteed to run by this worker. But in the worker.Initialize stage, we need to guarantee the worker run shard sync if shouldInitiateLeaseSync() returns true. So this syncShardsOnce should not call runShardSync, instead, it should simply call metricsEmittingShardSyncTask

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/PeriodicShardSyncManager.java
Comment (ychunxue) [reply src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/PeriodicShardSyncManager.java]: If result == null, that means metricsEmittingShardSyncTask.call() was having some issues and didn't return the expected result, right? Why do we return a valid test result with no exception in this case?

**Review by ychunxue — CHANGES_REQUESTED** [review]

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/PeriodicShardSyncManager.java
Comment (micah-jaffe) [reply src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/PeriodicShardSyncManager.java]: That makes sense. But if any Worker can run `syncShardsOnce` on bootstrap, instead of just elected leaders, doesn't that leave us vulnerable to a large fleet of workers coming online at the same time and hammering the lease table?

Granted we have some sleeping in `shouldInitiateLeaseSync()` to minimize contention, but still I thought the point of leader election on bootstrap was to handle this case.

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/PeriodicShardSyncManager.java
Comment (micah-jaffe) [reply src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/PeriodicShardSyncManager.java]: Removed this code as part of changes in above comment

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/PeriodicShardSyncManager.java
Comment (micah-jaffe) [reply src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/PeriodicShardSyncManager.java]: Discussed offline - the jitter in `shouldInitiateLeaseSync()` is the "leader election" we are using here, since current leader election mechanism uses lease table to determine eligible leaders, and the lease table in this case is empty. Updated to call ShardSyncTask directly.

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java
Comment (ychunxue) [inline src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java]: When would config.getWorkerIdentifier() be empty? We used to not have this check. Is this some cases introduced by our change?

Comment (micah-jaffe) [reply src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java]: This is set by the customer, so it's possible for it to be null or empty (isEmpty checks both). The unit tests were breaking with a null workerId so I added this - not sure why this wasn't the case before, I'll look into this further.

https://github.com/awslabs/amazon-kinesis-client/blob/v1.x/src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java#L1166

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/KinesisShardSyncer.java
Comment (joshua-kim) [inline src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/KinesisShardSyncer.java]: Don't need this TODO, lease cleanup will be added to shutdown instead of shardsync

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java
Comment (joshua-kim) [inline src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java:99]: Should we expose these as configurable for the customer? I know it's not in 2.x but I'm wondering if this is something the customer might want to configure - say the customer's application can't tolerate a 30 second gap of delay for their application / or can tolerate a longer delay of processing.

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java
Comment (micah-jaffe) [reply src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java]: Discussed offline, will modify the unit tests rather than setting random workerId. PeriodicShardSyncManager will fail to initialize with null or empty workerId, but that's consistent with the current behavior.

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/KinesisShardSyncer.java
Comment (micah-jaffe) [reply src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/KinesisShardSyncer.java]: Will remove

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java
Comment (micah-jaffe) [reply src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java:99]: This is a good point, I was thinking about that. The concern is that we already expose an overwhelming amount of configs, and I'm not sure what the customer demand for tuning this would be. 

On the plus side, this would make the unit tests run a lot faster if this was in a config. We can sync up with Ashwin for a final decision, but for now will leave that change out of this PR to keep consistent with 2.x.

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java
Comment (joshua-kim) [inline src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java]: I think it's easier to read this if we just use Optional.orElseGet()

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java
Comment (joshua-kim) [inline src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java]: I don't think we should ever allow `null` as a value for this to begin with, is there a case where this is valid? If not I think we can move this to constructor so we don't need to always null check this.

Comment (micah-jaffe) [reply src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java]: The entire KCL 1.x uses nulls instead of Optionals. Not ideal, but I think it's better to stick with the existing convention for the purposes of this PR

Nulls are valid when the Worker is initialized with the Builder, which sets nulls for all members not explicitly specified

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java
Comment (micah-jaffe) [reply src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java]: See below

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java
Comment (joshua-kim) [reply src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/Worker.java:99]: Yeah I don't think we should block on this, this is in sync with 2.x, I think this should be user configurable so we can always add this as a follow-up task after this PR.

**Review by ychunxue — APPROVED** [review]
LGTM

**Review by joshua-kim — APPROVED** [review]
LGTM

**Pull request merged**
